### PR TITLE
fix(autonomous): validate reused event roots at the append and update service boundaries (#7483)

### LINKED ISSUE #7483: fix(autonomous): validate reused event roots at the append service boundary
## Context

Follow-up to #7481 (merged via #7482). A hardening commit that was pushed to
the PR branch AFTER the squash-merge never landed on `main`, so the extra
validation it adds is currently missing from the mainline.

The autonomous caller already validates the reused `event_id` upstream, so this
is not a correctness regression in the shipped feature - it is defence in depth
that closes the reused-event link channel for any other (non-autonomous) caller.

## Scope

- `WorkflowService`: before linking reused event condition ids
  (`stepInput.conditionIds`), assert every non-blank id is an AND/OR
  finding-event root on THIS workflow - on both the simulation append path
  (`doAppendChainedStep`) and the isolated scenario-mirror path
  (`doAppendChainedStepToScenarioIsolated`). The link channel itself only checks
  root-ness, so this enforces the full event invariant at the service boundary.
- New `assertEventRootsOnWorkflow(workflowId, ids)` helper (no-op for
  null/empty).
- `WorkflowServiceTest`: a reused id that is not an event root is rejected at
  the boundary before the step-service link is reached.

### PR DESCRIPTION
**Pull request by SamuelHassine**
## Summary

- Recovers a hardening commit that was pushed to the #7481 / #7482 PR branch (`feat/autonomous-event-reuse`, `a7c3ecf46`) AFTER its squash-merge and therefore never landed on `main`, and extends it to the update channel found during review.
- `WorkflowService`: before linking reused event condition ids (`stepInput.conditionIds`), assert every non-blank id is an AND/OR finding-event root on the step's own workflow - on the simulation append path (`doAppendChainedStep`), the scenario append path (`doAppendChainedStepToScenario`, shared by the author-scenario overloads and the isolated REQUIRES_NEW mirror variants), and - added in review - the update path (`doUpdateChainedStep`), which linked reused ids through the same root-ness-only link channel (`linkExistingConditionsToStep` -> `findConditionRootById`). The link channel only checks existence + root-ness, so this enforces the full event invariant (root-ness AND AND/OR type AND same-workflow) at the service boundary.
- New `assertEventRootsOnWorkflow(workflowId, ids)` helper (no-op for null/empty). On the update side the step lookup is lazy - it only happens when a reused id is actually present, so the common no-reuse update pays nothing.
- Defence in depth only, no behaviour change for existing flows. Caller inventory: `AutonomousRunService` is the ONLY production caller reaching these boundaries with non-empty ids (no REST controller, importer, or copy flow does), and every site either pre-validates upstream (`assertEventRootOnSimulationWorkflow` / `assertEventRootOnScenarioWorkflow`) or passes mirror-recorded twin ids, where a stale entry now throws and is swallowed by the best-effort mirror exactly like a missing id already was. The manual UI's `condition_ids` channel goes through `StepApi` -> `StepService` directly and is a generic condition-root linking channel (not event reuse), so it is intentionally untouched.

Closes #7483

## Test plan

- [x] `WorkflowServiceTest` - 92 tests green, incl. 4 boundary tests: the simulation append path rejects a non-root reused id, the simulation append path rejects a cross-workflow event root, the isolated scenario-mirror append path rejects a non-root reused id, and the update path rejects a non-root reused id before any rebuild (missing-id and non-event-type rejections were already covered via the singular assert tests from #7482)
- [x] `spotless:check` clean
- [x] CI green

### CODE REVIEW

File: openaev-api/src/main/java/io/openaev/service/chaining/WorkflowService.java [hunk 1]
```diff
@@ -1969,6 +1981,25 @@ public void assertEventRootOnScenarioWorkflow(String scenarioId, String eventId)
...
+      throws ChainingException {
+    if (eventConditionIds == null) {
+      return;
+    }
+    for (String eventConditionId : eventConditionIds) {
+      if (hasText(eventConditionId)) {
+        assertEventRootOnWorkflow(workflowId, eventConditionId);
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply openaev-api/src/main/java/io/openaev/service/chaining/WorkflowService.java:2001]: Deliberately keeping the per-id loop. Every production caller of this boundary passes at most ONE id: the autonomous channel is `List.of(requestedEventId)` on all append/update sites (the caller inventory is in the PR description), and the manual UI's multi-root linking goes through `StepApi` -> `StepService` directly, never through this service boundary. The path is LLM-paced (a handful of authored steps per run), and the loop preserves the singular helper's precise per-id error (missing vs child vs non-event type vs cross-workflow) - a bulk `findAllById` would add repository surface and still need a per-id pass to produce those messages. Happy to revisit with a bulk fetch if this channel ever carries operator-driven bulk lists.

File: openaev-api/src/test/java/io/openaev/service/chaining/WorkflowServiceTest.java [hunk 2]
```diff
@@ -2564,5 +2565,28 @@ void given_aNonRoot_when_resolvingInputs_then_itIsEmpty() {
...
+          .thenReturn(condition("leaf", ConditionType.EQ, root, RUN_WORKFLOW_ID));
+
+      assertThrows(
+          ChainingException.class,
+          () ->
+              workflowService.appendChainedStep(
+                  SIMULATION_ID, new InjectInput(), null, List.of(), List.of("leaf")));
+      // Rejected at the boundary: the step-service link is never reached.
+      verifyNoInteractions(stepService);
+    }
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply openaev-api/src/test/java/io/openaev/service/chaining/WorkflowServiceTest.java:2590]: Done in ee6c5d226: added `given_aNonRootReusedId_when_appendingToScenarioIsolated_then_itThrowsBeforeLinking`, exercising `appendChainedStepToScenarioIsolated` - the REQUIRES_NEW entry the scenario mirror actually uses; it shares `doAppendChainedStepToScenario` with `appendChainedStepToScenario`, so both public entries are locked in. The same commit also adds a cross-workflow rejection test on the simulation append boundary and - found while reviewing the surroundings - extends the same boundary assert to the UPDATE channel (`doUpdateChainedStep`), which linked reused ids through the same root-ness-only link channel; covered by `given_aNonRootReusedId_when_updating_then_itThrowsBeforeRebuilding`.

**Comment by SamuelHassine** [comment]
Review-and-fix summary (independent senior pass + Copilot thread triage)

Verdict: APPROVE on content. GitHub refuses self-approval on one's own PR, so this stands as the review verdict - a maintainer approval is still required by branch protection.

Scope reviewed beyond the diff: both append bodies end-to-end, the singular assertEventRootOnWorkflow (all four rejection reasons verified: missing id, child condition, non-AND/OR type, cross-workflow root), EVENT_ROOT_TYPES, the whole conditionIds link channel down through StepService/ConditionService (findConditionRootById checks existence + root-ness ONLY, which is exactly what makes this boundary assert meaningful), the update overloads added by #7482, and a complete production caller inventory.

Findings and dispositions:

- Caller compatibility (the critical check): AutonomousRunService is the ONLY production caller reaching appendChainedStep* / appendChainedStepToScenario* / updateChainedStep* with non-empty reused ids - no REST controller, importer, or copy/duplicate flow uses this channel. Every call site either pre-validates upstream (assertEventRootOnSimulationWorkflow / assertEventRootOnScenarioWorkflow) or passes mirror-recorded twin ids, where a stale entry now throws and is swallowed by the best-effort mirror exactly like a missing id already was. The assert cannot 400 any existing working flow. The manual UI's condition_ids channel (StepApi -> StepService) is a GENERIC condition-root linking channel, not event reuse, so it is deliberately out of scope.
- Update-path gap (must-fix, found in review, fixed in ee6c5d226): doUpdateChainedStep linked caller-supplied reused event ids through the same root-ness-only link channel WITHOUT the boundary assert - the exact hole this PR closes for append, on the update channel. Now asserted there too, with a lazy step lookup that only runs when a reused id is actually present.
- Copilot thread 1 (N+1 bulk fetch): declined with justification - every production caller passes at most ONE id (List.of(requestedEventId)), the path is LLM-paced, and the per-id loop preserves the precise per-id error messages; a bulk findAllById would still need a per-id pass to produce them.
- Copilot thread 2 (scenario-twin test): implemented, plus a cross-workflow rejection test on the sim append boundary and an update-boundary test. Missing-id and non-event-type rejections were already covered by the #7482 singular-assert tests, so they were not duplicated.

State: CI fully green on ee6c5d226 (all pipeline jobs incl. API tests, E2E, CodeQL, spotless), 0 unresolved review threads (post-push Copilot re-review generated no new comments), mergeable CLEAN. WorkflowServiceTest: 92 tests green. Title and body updated to reflect the widened scope (append + update boundaries).

Stale branch cleanup: the resurrected feat/autonomous-event-reuse remote branch is already deleted; content parity was verified first (diff of a7c3ecf46 against this PR's final tree for the two touched files shows the recovered commit reproduced the stranded hardening exactly - the only delta is this review's own additions).

Remaining blockers: none in the PR itself - just the required maintainer approval.
